pkg/kubectl/deployment.go

```diff
@@ -56,18 +74,37 @@ func (s DeploymentBasicGeneratorV1) Generate(params map[string]interface{}) (run
 	if !isArray {
 		return nil, fmt.Errorf("expected []string, found :%v", imageStrings)
 	}
-	delegate := &DeploymentBasicGeneratorV1{Name: name, Images: imageStrings}
-	return delegate.StructuredGenerate()
+	return &BaseDeploymentGenerator{
+		Name:   name,
+		Images: imageStrings,
+	}, nil
 }
 
-// StructuredGenerate outputs a deployment object using the configured fields
-func (s *DeploymentBasicGeneratorV1) StructuredGenerate() (runtime.Object, error) {
-	if err := s.validate(); err != nil {
-		return nil, err
+// structuredGenerate: determine the fields of a deployment. The struct that
+// embeds BaseDeploymentGenerator should assemble these pieces into a
+// runtime.Object.
+func (b BaseDeploymentGenerator) structuredGenerate() (
+	podSpec v1.PodSpec,
+	labels map[string]string,
+	selector metav1.LabelSelector,
+	err error,
+) {
+	err = b.validate()
+	if err != nil {
+		return
 	}
+	podSpec = buildPodSpec(b.Images)
+	labels = map[string]string{}
+	labels["app"] = b.Name
+	selector = metav1.LabelSelector{MatchLabels: labels}
+	return
+}
 
+// buildPodSpec: parse the image strings and assemble them into the Containers
+// of a PodSpec. This is all you need to create the PodSpec for a deployment.
+func buildPodSpec(images []string) v1.PodSpec {
 	podSpec := v1.PodSpec{Containers: []v1.Container{}}
-	for _, imageString := range s.Images {
+	for _, imageString := range images {
```

Comment:
Ah nevermind, figured it out.  This is just for the name.  While you are here, consider pulling this into a function: `getContainerName(image string) string`.  Also, we should check that there are not 2 instance of the same container image running.  I would expect this to be rare, but it would result in weird things.  Maybe keep a set of names we have seen and throw an error?